### DISCUSSION TITLE
Offline vulnerability detection

### DISCUSSION THREAD

**Discussion by MathisLandreau** [wazuh/wazuh#27194] [Q&A] [answered]
I'm currently testing wazuh. In this context, I did the Vulnerability detection test in offline mode and I have a very big difference in the number of CVEs detected between the two. In offline mode, I have more than half the number of CVEs detected. How can this be explained or solved? I use the same agent and the wazuh offline documentation for the zip to be recovered.   

e.g. vim-common 2:9.0.1378-2 detects CVE-2023-1355 but in hore line it is not detected.

thank you for your help.

**Comment by Damian-Mangold** [comment, marked as answer]
Hello @MathisLandreau. First, let me explain the differences in vulnerability content between the **online** and **offline** update modes of the manager:  

**Online Update Mode:**  
In this mode, the manager automatically checks for updates every hour. When new updates are detected, they are downloaded and immediately incorporated into the database. This ensures that the vulnerability content remains as up-to-date as possible.  

**Offline Update Mode:**  
This mode relies on a pre-generated file containing vulnerability information. This file is currently updated once a week, on Mondays. Until this file is updated and loaded into the manager, it will not reflect the latest vulnerabilities.  

The discrepancy in detections between a manager configured in online mode and one in offline mode is likely due to these differences in update mechanisms. The online mode provides frequent updates, while the offline mode inherently experiences a delay due to the weekly file generation and distribution process.  

---

**Recommendation:**  
I suggest waiting until Tuesday, when the new offline content becomes available. Once you have updated the manager in offline mode, please check the detections again. If the differences persist, I kindly ask that you share a detailed list of the vulnerabilities detected by both managers so I can perform a comparative analysis and identify the reasons for the discrepancy.  

Feel free to reach out if you have any questions.

**Reply by MathisLandreau** [reply]
the problem is that it's not just the latest CVEs that are missing. I'll give you the list on a test machine. 

[events-2024-12-06T13_03_49.760Z.csv](https://github.com/user-attachments/files/18107379/events-2024-12-06T13_03_49.760Z.csv)
[events-2024-12-06T12_45_37.864Z.csv](https://github.com/user-attachments/files/18107380/events-2024-12-06T12_45_37.864Z.csv)
thank you for your help.

**Reply by Damian-Mangold** [reply]
Hello @MathisLandreau. 

> the problem is that it's not just the latest CVEs that are missing.

The vulnerability content is constantly being improved. Therefore, not only new CVEs are updated, but also old ones.

> I'll give you the list on a test machine.

Could you provide more details about the files you shared with me? This will help me better understand the situation and proceed with the analysis.

Have you updated the offline manager with the latest version of the content published this Monday? It’s important to confirm this step, as it could be related to the differences observed.


---

To conduct a more detailed analysis, I will need the following information:

- [OS information](https://documentation.wazuh.com/current/user-manual/api/reference.html#operation/api.controllers.syscollector_controller.get_os_info)
- [Package inventory](https://documentation.wazuh.com/current/user-manual/api/reference.html#operation/api.controllers.syscollector_controller.get_packages_info)
- Vulnerabilities inventory (Manager online)
- Vulnerabilities inventory (Manager offline)


If you need help to obtain all the information, please consult [here](https://github.com/wazuh/wazuh/issues/25001#issuecomment-2275709492).

**Reply by MathisLandreau** [reply]
hello, 
the two files are a vulnerability detection test on the same machine but one is in “normal” mode and the other in offline mode. the first is in normal mode and the other in offline mode. the problem is that they are all identical with one test machine. on one side i have 514 vulnerabilities found and on the other 63... the difference seems big to me. 

example : vim-runtime CVE-2023-48234 she is detected in "normal" mode but on offline mode no.  I check the vulnerbyliti in the json file and it's there.  

/var/lib/dpkg/status: 
Package: vim-runtime
Status: install ok installed
Priority: optional
Section: editors
Installed-Size: 36406
Maintainer: Debian Vim Maintainers <team+vim@tracker.debian.org>
Architecture: all
Multi-Arch: foreign
Source: vim
Version: 2:9.0.1378-2
Recommends: vim | vim-gtk3 | vim-motif | vim-nox | vim-tiny
Breaks: vim-tiny (<< 2:9.0.1378-2)
Enhances: vim-tiny
Description: Vi IMproved - Runtime files

With your recomendation I've looked but no error comes up.

I add a question: is it possible to have the structure of a CVE in json? to create one.  

thank you.

**Reply by Damian-Mangold** [reply]
Thank you very much for the clarification. Let me analyze the information shared.

**Reply by Damian-Mangold** [reply]
Hi @MathisLandreau . In order to perform some tests, could you please provide me with the following information from the agents where you're observing the issue?

- [OS information](https://documentation.wazuh.com/current/user-manual/api/reference.html#operation/api.controllers.syscollector_controller.get_os_info)

If you need help to obtain all the information, please consult [here](https://github.com/wazuh/wazuh/issues/25001#issuecomment-2275709492).

Thank you!

**Reply by MathisLandreau** [reply]
Hello, I'm still testing with a single agent. 
os : 
[OS.txt](https://github.com/user-attachments/files/18127257/OS.txt)
packages: 
[Packages.txt](https://github.com/user-attachments/files/18127258/Packages.txt)
hotfixes: 
[Hotfixes.txt](https://github.com/user-attachments/files/18127260/Hotfixes.txt)

**Reply by Damian-Mangold** [reply]
Hi @MathisLandreau. Could you share with me the databases `/var/ossec/queue/vd_updater/` of both managers to verify the version of content they have downloaded?

**Reply by MathisLandreau** [reply]
yes, here are the contents:  
[vd_updater.zip](https://github.com/user-attachments/files/18127885/vd_updater.zip)

I just removed the json file because it was too heavy . but I download the latest version from 2024-12-12T16:02:15.678097Z

**Reply by Damian-Mangold** [reply]
Hi @MathisLandreau. Indeed, the vulnerabilities database is significantly outdated, which explains the discrepancy in detections. Your database is currently on version 851035, while the latest version is 1128987.

We need to identify the reason why the database is not being updated correctly.  To analyze in more detail what might be happening, I kindly ask you to enable debug logging mode on the manager. Please follow these steps:

1- Open the configuration file using the following command: `nano /var/ossec/etc/internal_options.conf`
2- Set the logging level by adding or modifying the following option: `wazuh_modules.debug=2`
3- Restart the manager service to apply the changes: `systemctl restart wazuh-manager`
4- Wait a few minutes (30min) for the manager to fully start.
5- Once ready, please share the complete log file located at: `/var/ossec/logs/ossec.log`

This file will allow us to verify if there are any errors related to the vulnerability detector and proceed with a detailed analysis.

**Reply by MathisLandreau** [reply]
> Hi @MathisLandreau. Indeed, the vulnerabilities database is significantly outdated, which explains the discrepancy in detections. Your database is currently on version 851035, while the latest version is 1128987.
> 
> 
> 
> We need to identify the reason why the database is not being updated correctly.  To analyze in more detail what might be happening, I kindly ask you to enable debug logging mode on the manager. Please follow these steps:
> 
> 
> 
> 1- Open the configuration file using the following command: `nano /var/ossec/etc/internal_options.conf`
> 
> 2- Set the logging level by adding or modifying the following option: `wazuh_modules.debug=2`
> 
> 3- Restart the manager service to apply the changes: `systemctl restart wazuh-manager`
> 
> 4- Wait a few minutes (30min) for the manager to fully start.
> 
> 5- Once ready, please share the complete log file located at: `/var/ossec/logs/ossec.log`
> 
> 
> 
> This file will allow us to verify if there are any errors related to the vulnerability detector and proceed with a detailed analysis.

Hello, thanks for the help, I'm doing this on Thursday morning. I'll get back to you with the results.

**Reply by MathisLandreau** [reply]
Hello, here is the file you need: 
[ossec.log](https://github.com/user-attachments/files/18194591/ossec.log)

**Reply by Damian-Mangold** [reply]
Hi @MathisLandreau. After analyzing the log, I believe I’ve identified the issue. Your offline mode is currently configured to use the following file:  

```bash
"offline-url":"file:///var/ossec/queue/vulnerabilities/1117976_1733137820.zip"
```  
However, when the manager attempts to process it, the file cannot be found. Below is a log snippet where this behavior is observed:  

```
2024/12/19 07:30:40 wazuh-modulesd:vulnerability-scanner[856] wm_vulnerability_scanner.c:45 at wm_vulnerability_scanner_log_config(): DEBUG: {"vulnerability-detection":{"enabled":"yes","index-status":"yes","feed-update-interval":"60m","offline-url":"file:///var/ossec/queue/vulnerabilities/1117976_1733137820.zip","cti-url":"https://cti.wazuh.com/api/v1/catalog/contexts/vd_1.0.0/consumers/vd_4.8.0"},"wmMaxEps":100,"translationLRUSize":2048,"osdataLRUSize":1000,"remediationLRUSize":2048,"managerDisabledScan":1,"indexer":{"enabled":"yes","hosts":["https://wazuh.indexer:9200"],"ssl":{"certificate_authorities":["/etc/ssl/root-ca.pem"],"certificate":"/etc/ssl/filebeat.pem","key":"/etc/ssl/filebeat.key"}},"clusterEnabled":false,"clusterName":"wazuh.manager","clusterNodeName":"undefined"}

2024/12/19 07:30:41 wazuh-modulesd:content-updater[856] offlineDownloader.hpp:61 at copyFile(): WARNING: File 'file:///var/ossec/queue/vulnerabilities/1117976_1733137820.zip' doesn't exist.
```  

When updating the offline content, you have two options:  
1. **Update the file path in the manager’s configuration file (`ossec.conf`)**: Modify the name or path of the new file in the configuration and restart the manager service to apply the changes.  
2. **Replace the existing file**: Overwrite the previous file with the new one, ensuring they have the exact same name.  

Please verify that the full path specified in the `ossec.conf` file matches the exact location and name of the update file. Once the paths and names are aligned, the manager should update correctly.  After aligning it, restart the manager to force the update. If you want to wait a few minutes and share the log file with me again to validate that everything works.

**Reply by Damian-Mangold** [reply]
Hi @MathisLandreau. I see that after that error you updated the file path and the manager was updated.

```shell
2024/12/19 07:35:59 wazuh-modulesd:vulnerability-scanner[2934] wm_vulnerability_scanner.c:45 at wm_vulnerability_scanner_log_config(): DEBUG: {"vulnerability-detection":{"enabled":"yes","index-status":"yes","feed-update-interval":"60m","offline-url":"file:///var/ossec/queue/vulnerabilities/1131217_1734339092.zip","cti-url":"https://cti.wazuh.com/api/v1/catalog/contexts/vd_1.0.0/consumers/vd_4.8.0"},"wmMaxEps":100,"translationLRUSize":2048,"osdataLRUSize":1000,"remediationLRUSize":2048,"managerDisabledScan":1,"indexer":{"enabled":"yes","hosts":["https://wazuh.indexer:9200"],"ssl":{"certificate_authorities":["/etc/ssl/root-ca.pem"],"certificate":"/etc/ssl/filebeat.pem","key":"/etc/ssl/filebeat.key"}},"clusterEnabled":false,"clusterName":"wazuh.manager","clusterNodeName":"undefined"}

2024/12/19 07:50:37 wazuh-modulesd:vulnerability-scanner[2934] databaseFeedManager.hpp:349 at operator()(): INFO: Feed update process completed.
```

Could you share with me again the database (`/var/ossec/queue/vd_updater/`) to verify the feed version.

**Reply by MathisLandreau** [reply]
yes it doesn't find it that's normal because it was the moment just before I updated the zip so I deleted the existing one with the new one which starts at 07:36:18.

No problem, I'll pass it on to you: 
[vd_updater.zip](https://github.com/user-attachments/files/18197069/vd_updater.zip)

**Reply by Damian-Mangold** [reply]
The vulnerabilities content is now updated to the latest available version. I recommend waiting for the offline manager to reanalyze the agent. Once the analysis is complete, the differences in detections between the offline and online managers should be minimal or nonexistent.

If you notice any significant discrepancies after the analysis, please feel free to share the details so I can perform a more thorough review.

**Reply by MathisLandreau** [reply]
I have just realized : 
       -offline  [events-2024-12-19T14_47_11.113Z.csv](https://github.com/user-attachments/files/18199543/events-2024-12-19T14_47_11.113Z.csv)
       -'Normal' :  [events-2024-12-19T14_58_45.816Z.csv](https://github.com/user-attachments/files/18199546/events-2024-12-19T14_58_45.816Z.csv)

now 62 offline and 513 normal. I think the difference is still big. I'd like to hear your opinion.

I'll try to get the latest version and start again with docker.

**Reply by Damian-Mangold** [reply]
There is definitely a lot of difference. Could you share with me now the ossec.log of the online (normal) manager to analyze if there is any error.

**Reply by MathisLandreau** [reply]
Hello, 
I updated Wazuh to the latest version (4.9.2) and it became the same for offline. so I do not know where the problem came from but once all updated all is good.
Thank you for your help and your time.

**Reply by Damian-Mangold** [reply]
Great news !!!. Surely there was some problem during the update which caused the differences in detection.

If you have any questions please do not hesitate to contact us again.

Best regards.
